react/features/calendar-sync/middleware.js

```diff
@@ -91,64 +145,81 @@ function _ensureDefaultServer(store) {
-                            logger.warn(
-                                'Skipping calendar event due to invalid dates',
-                                event.title,
-                                event.startDate,
-                                event.endDate
-                            );
-                        } else if (eventEndDate > now) {
-                            eventList.push({
-                                endDate: eventEndDate,
-                                id: event.id,
-                                startDate: eventStartDate,
-                                title: event.title,
-                                url: jitsiURL
-                            });
+function _fetchCalendarEntries(
+        store,
+        maybePromptForPermission,
+        forcePermission
+) {
+    const { dispatch } = store;
+    const state = store.getState()['features/calendar-sync'];
+    const { calendarAccessStatus } = state;
+    const promptForPermission
+        = (maybePromptForPermission && !calendarAccessStatus)
+        || forcePermission;
+
+    _ensureCalendarAccess(promptForPermission, dispatch)
+    .then(accessGranted => {
+        if (accessGranted) {
+            const startDate = new Date();
+            const endDate = new Date();
+
+            startDate.setDate(startDate.getDate() + FETCH_START_DAYS);
+            endDate.setDate(endDate.getDate() + FETCH_END_DAYS);
+
+            RNCalendarEvents.fetchAllEvents(
+                startDate.getTime(),
+                endDate.getTime(),
+                []
+            )
+            .then(events => {
+                const { knownDomains } = state;
+                const eventList = [];
+
+                if (events && events.length) {
+                    for (const event of events) {
+                        const jitsiURL = _getURLFromEvent(event, knownDomains);
+                        const now = Date.now();
+
+                        if (jitsiURL) {
+                            const eventStartDate = Date.parse(event.startDate);
+                            const eventEndDate = Date.parse(event.endDate);
+
+                            if (isNaN(eventStartDate) || isNaN(eventEndDate)) {
+                                logger.warn(
+                                    'Skipping invalid calendar event',
+                                    event.title,
+                                    event.startDate,
+                                    event.endDate
+                                );
+                            } else if (eventEndDate > now) {
+                                eventList.push({
+                                    endDate: eventEndDate,
+                                    id: event.id,
+                                    startDate: eventStartDate,
+                                    title: event.title,
+                                    url: jitsiURL
+                                });
+                            }
```

Comment:
This is a pretty nice pyramid. Can you think of a way to reduce its girth a bit? If you can't, you can't. It's not that complicated anyway, but maybe even a named helper would make it clear in one glance what this block does.

Reply:
Wat? You no like pyramids? :)

Refactored, thanks!